### PR TITLE
New architecture prototype

### PR DESCRIPTION
**Pull request by kolesnikovae**
The PR is the main development branch for the work on the new architecture (https://github.com/grafana/pyroscope/issues/3192). ~The PR will never be merged; it was created only for visibility. Once the prototype is finished, we will publish appropriate documentation based on the experiment results.~

---

Update:

The experiment is over. The results are very promising.

The new code is very well isolated and does not affect behaviour of the current version, therefore I'd like to merge the changes: we will continue working on the new components in the main branch.

### CODE REVIEW

File: pkg/metastore/metastore_compaction_queue.go [hunk 1]
```diff
@@ -84,50 +78,45 @@ func (c *jobQueueEntry) load(job *compactionpb.CompactionJob) {
...
-		deadline: 0, // Deadline and token will be assigned
-		token:    0, // when the job is "dequeued" (transferred).
-		proto:    job,
+		status: js,
+		index:  0,
+		proto:  job,
 	}
+	// Deadline will be assigned when the job is "dequeued" (transferred).
+	job.RaftLogIndex = 0
+	job.LeaseExpiresAt = 0
```
Comment (kolesnikovae) [inline pkg/metastore/metastore_compaction_queue.go:87]: This is only true for job in the initial state. If the job already has index we do want to restore this information

File: pkg/phlaredb/tsdb/index/index.go [hunk 2]
```diff
@@ -279,14 +283,14 @@ type FileWriter struct {
 	name string
 }
 
-func NewFileWriter(name string) (*FileWriter, error) {
+func NewFileWriter(name string, bufferSize int) (*FileWriter, error) {
```
Comment (korniltsev) [inline pkg/phlaredb/tsdb/index/index.go:286]: I will revert these changes in tsdb/index/index.go - it is not used anymore

File: api/metastore/v1/metastore.proto [hunk 3]
```diff
@@ -0,0 +1,71 @@
+syntax = "proto3";
+
+package metastore.v1;
```
Comment (korniltsev) [inline api/metastore/v1/metastore.proto:3]: nit: maybe we should put it in v2 or v2-experimental package

File: pkg/phlaredb/tsdb/index/index.go (hunk 2 above)
Comment (kolesnikovae) [reply pkg/phlaredb/tsdb/index/index.go:286]: It's currently used in compactor. I see no need to change it now

File: api/metastore/v1/metastore.proto (hunk 3 above)
Comment (kolesnikovae) [reply api/metastore/v1/metastore.proto:3]: This is the first version of metastore API.

As long as the API is not public, I see no practical benefits of separating the API

**Review by korniltsev — APPROVED** [review]
lgtm

**Review by simonswine — APPROVED** [review]
Thank you so much for cleaning that up and make it mergable it main @kolesnikovae .

Also kudos to everyone else working on this experiment 👏

**Pull request merged**